lib/mps_statespace.h

```diff
@@ -232,6 +233,149 @@ class MPSStateSpace {
     return partial_contract(0, 0);
   }
 
+  // Compute the 2x2 1-RDM of state on index. Result written to rdm.
+  // Requires: scratch to be allocated.
+  static void ReduceDensityMatrix(MPS& state, MPS& scratch, int index,
+                                  fp_type* rdm) {
+    const auto num_qubits = state.num_qubits();
+    const auto bond_dim = state.bond_dim();
+    const auto end = Size(state);
+    const bool last_index = (index == num_qubits - 1);
+    const auto right_dim = (last_index ? 1 : bond_dim);
+    auto offset = 0;
+    fp_type* state_raw = state.get();
+    fp_type* scratch_raw = scratch.get();
+
+    Copy(state, scratch);
+
+    // Contract leftmost blocks together, store result in state scratch.
+    ConstMatrixMap top((Complex*)scratch_raw, 2, bond_dim);
+    ConstMatrixMap bot((Complex*)state_raw, 2, bond_dim);
+    MatrixMap partial_contract((Complex*)(state_raw + end), bond_dim, bond_dim);
+    MatrixMap partial_contract2(
+        (Complex*)(state_raw + end + 2 * bond_dim * bond_dim), bond_dim,
+        2 * bond_dim);
+
+    partial_contract.setZero();
+    partial_contract(0, 0) = 1;
+    if (index > 0) {
+      partial_contract.noalias() = top.adjoint() * bot;
+    }
+
+    // Contract all internal blocks together.
+    for (unsigned i = 1; i < index; ++i) {
+      offset = GetBlockOffset(state, i);
+
+      // reshape:
+      new (&partial_contract2)
+          MatrixMap((Complex*)(state_raw + end + 2 * bond_dim * bond_dim),
+                    bond_dim, 2 * bond_dim);
+
+      // Merge bot into left boundary merged tensor.
+      new (&bot) ConstMatrixMap((Complex*)(state_raw + offset), bond_dim,
+                                2 * bond_dim);
+      partial_contract2.noalias() = partial_contract * bot;
+
+      // reshape:
+      new (&partial_contract2)
+          MatrixMap((Complex*)(state_raw + end + 2 * bond_dim * bond_dim),
+                    2 * bond_dim, bond_dim);
+
+      // Merge top into partial_contract2.
+      new (&top) ConstMatrixMap((Complex*)(scratch_raw + offset), 2 * bond_dim,
+                                bond_dim);
+      partial_contract.noalias() = top.adjoint() * partial_contract2;
+    }
+
+    // The [bond_dim, bond_dim] block in state_raw now contains the contraction
+    // up to, but not including index.
+    // Contract rightmost blocks.
+    offset = GetBlockOffset(state, num_qubits - 1);
+    new (&top) ConstMatrixMap((Complex*)(scratch_raw + offset), bond_dim, 2);
+    new (&bot) ConstMatrixMap((Complex*)(state_raw + offset), bond_dim, 2);
+    new (&partial_contract)
+        MatrixMap((Complex*)(scratch_raw + end), bond_dim, bond_dim);
+    new (&partial_contract2)
+        MatrixMap((Complex*)(scratch_raw + end + 2 * bond_dim * bond_dim),
```

Comment:
Nit: set a variable to `(Complex*)(scratch_raw + end + 2 * bond_dim * bond_dim)`? This expression is used four times in this function.